accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+  smart() => smartHelper(this);  // valid
+  static smartHelper(Object o) { ... }
+}
+...
+  MySmart.smartHelper(someObject);  // valid
+```
+
+### Semantics of Invocations
+
+If an extension is found to be the one applying to a member invocation, then at run-time, the invocation will perform a method invocation of the corresponding instance member of the extension, with `this` bound to the receiver value and type parameters bound to the types found by static inference.
+
+If the receiver is `null`, then that invocation is an immediate run-time error unless the `on` type of the extension has a trailing `?`.
+
+With NNBD types, it is a run-time error if the receiver is `null` and the instantiated `on` type of the selected extension does not allow `null`. For sound non-nullable types, this can be excluded statically. If a receiver expression is nullable, then the matching `on` type must be nullable too. If a receiver expression is soundly non-nullable, then the value cannot be `null` at run-time. During migration, we will have unsound nullable types like `int*` which should be matched by `extension Wot on int {…}`, and in that case we will need a run-time check to avoid passing `null` to that extension.
+
+### Semantics of Extension Members
+
+When executing an extension instance member, we stated earlier that the member is invoked with the original receiver as `this` object. We still have to describe how that works, and what the lexical scope is for those members.
+
+Inside an extension declaration, we give preference to that extension over any other extension. At any otherwise invalid invocation, if the current extension applies, then it is considered more specific than all other extensions that may apply. This applies to static methods too. 
+
+Example:
+
+```dart
+extension MyUnaryNumber on List<Object> {
+  bool get isEven => this.length.isEven;
+  bool get isOdd => !this.isEven;
+  static bool isListEven(List<Object> list) => list.isEven;
+}
+```
+
+Here the `list.isEven` is  guaranteed to hit the `isEven` of the same extension (unless someone puts an `isEven` member on `List`), an extension has more specificity than any other extension inside itself.
+
+About *`this`*: Inside an instance member of an extension, the static type of the `this` operator is the `on` type. However, for member invocations, we give preference to extension methods of the current extension, *even over members of the `on` type*. 
+
+Here, the `this.isEven` is a member invocation on `this`, so it *first* tries to resolve `isEven` against the current extension. If it finds a member with the same base name on the extension, that member is invoked with the same `this` value. If not, then it is considered a normal member invocation on an object with static type `List<Object>,` and if `List<Object>` does not have the necessary member, then all applicable extensions are checked (which will definitely not match the current extension since we already checked that).
```

Comment:
I'm not sure what to think about this.  It feels a bit odd to resolve these differently from external uses, but it also feels a bit odd for unqualified references to resolve to the instance method.